components/servo/message_port.rs

```diff
@@ -0,0 +1,163 @@
+
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use base::id::{MessagePortId, MessagePortRouterId, WebViewId};
+use constellation_traits::{EmbedderToConstellationMessage, MessagePortMsg, PostMessageData};
+use embedder_traits::{EmbedderMsg, EmbedderProxy, JSValue};
+use ipc_channel::ipc;
+use ipc_channel::router::ROUTER;
+
+use crate::{ConstellationProxy, WebView};
+
+pub trait MessagePortDelegate {
+    /// Invoked when a message is posted on an entangled MessagePort. The
+    /// posted data is provided as an argument, and only includes seralized
+    /// versions of fundamental JS data types.
+    fn post_message_received(&self, _webview: WebView, _message_port: MessagePort, _data: JSValue) {
+    }
+}
+
+enum MessagePortState {
+    Ready {
+        id: MessagePortId,
+        webview_id: WebViewId,
+        entangled_id: MessagePortId,
+        constellation_proxy: ConstellationProxy,
+        delegate: Option<Rc<dyn MessagePortDelegate>>,
+    },
+    Detached,
+}
+
+#[derive(Clone)]
+pub struct MessagePort(Rc<RefCell<MessagePortState>>);
+
+#[derive(Debug, PartialEq)]
+pub struct MessagePortDetached;
+
+impl MessagePort {
+    fn new(
+        constellation_proxy: ConstellationProxy,
+        webview_id: WebViewId,
+        port_id: MessagePortId,
+        entangled_id: MessagePortId,
+    ) -> MessagePort {
+        MessagePort(Rc::new(RefCell::new(MessagePortState::Ready {
+            constellation_proxy,
+            webview_id,
+            id: port_id,
+            entangled_id,
+            delegate: None,
+        })))
+    }
+
+    pub(crate) fn new_entangled(
+        constellation_proxy: ConstellationProxy,
+        embedder_proxy: EmbedderProxy,
+        webview_id: WebViewId,
+    ) -> (MessagePort, MessagePort) {
+        let (port_control_sender, port_control_receiver) =
+            ipc::channel().expect("ipc channel failure");
+        ROUTER.add_typed_route(
+            port_control_receiver,
+            Box::new(move |message| match message.unwrap() {
+                MessagePortMsg::CompleteTransfer(..) => unreachable!(),
+                MessagePortMsg::CompletePendingTransfer(..) => unreachable!(),
+                MessagePortMsg::CompleteDisentanglement(..) => unreachable!(),
+                MessagePortMsg::NewTask(message_port_id, task) => {
+                    let data = match task.data {
+                        PostMessageData::StructuredClone(..) => unreachable!(),
+                        PostMessageData::Serialized(data) => data,
+                    };
+                    let _ = embedder_proxy.send(EmbedderMsg::PostMessage(
+                        webview_id,
+                        message_port_id,
+                        data,
+                    ));
+                },
+            }),
+        );
+        let first_id = MessagePortId::new();
+        let second_id = MessagePortId::new();
+        let router_id = MessagePortRouterId::new();
```

Comment:
I think you should use only one router for all ports used by the embedder, but that requires quite some changes to this module.